crates/xilem_html/src/vecmap.rs

```diff
@@ -0,0 +1,382 @@
+    ///
+    /// map.insert(37, "b");
+    /// assert_eq!(map.insert(37, "c"), Some("b"));
+    /// assert_eq!(map[&37], "c");
+    /// ```
+    pub fn insert(&mut self, key: K, value: V) -> Option<V>
+    where
+        K: Ord,
+    {
+        match self.0.binary_search_by_key(&&key, |(n, _)| n) {
+            Ok(pos) => {
+                let mut val = (key, value);
+                std::mem::swap(&mut self.0[pos], &mut val);
+                Some(val.1)
+            }
+            Err(pos) => {
+                self.0.insert(pos, (key, value));
+                None
+            }
+        }
+    }
+
+    /// Removes a key from the map, returning the value at the key if the key
+    /// was previously in the map.
+    ///
+    /// The key may be any borrowed form of the map's key type, but the ordering
+    /// on the borrowed form *must* match the ordering on the key type.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```ignore
+    /// # use crate::vecmap::VecMap;
+    /// let mut map = VecMap::default();
+    /// map.insert(1, "a");
+    /// assert_eq!(map.remove(&1), Some("a"));
+    /// assert_eq!(map.remove(&1), None);
+    /// ```
+    pub fn remove<Q: ?Sized>(&mut self, key: &Q) -> Option<V>
+    where
+        K: Borrow<Q> + Ord,
+        Q: Ord,
+    {
+        // TODO not sure whether just a simple find is better here? Probably needs more benching
+        match self.0.binary_search_by_key(&key, |(k, _)| k.borrow()) {
+            Ok(pos) => Some(self.0.remove(pos).1),
+            Err(_) => None,
+        }
+    }
+
+    /// Returns `true` if the map contains no elements.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```ignore
+    /// # use crate::vecmap::VecMap;
+    /// let mut a = VecMap::default();
+    /// assert!(a.is_empty());
+    /// a.insert(1, "a");
+    /// assert!(!a.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the number of elements in the map.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```ignore
+    /// # use crate::vecmap::VecMap;
+    /// let mut a = VecMap::default();
+    /// assert_eq!(a.len(), 0);
+    /// a.insert(1, "a");
+    /// assert_eq!(a.len(), 1);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl<K, Q: ?Sized, V> Index<&Q> for VecMap<K, V>
+where
+    K: Borrow<Q> + Ord,
+    Q: Ord,
+{
+    type Output = V;
+
+    /// Returns a reference to the value corresponding to the supplied key.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the key is not present in the `VecMap`.
+    #[inline]
+    fn index(&self, key: &Q) -> &V {
+        self.get(key).expect("no entry found for key")
+    }
+}
+
+impl<'a, K, V> IntoIterator for &'a VecMap<K, V> {
+    type Item = (&'a K, &'a V);
+
+    type IntoIter = std::iter::Map<std::slice::Iter<'a, (K, V)>, fn(&'a (K, V)) -> (&'a K, &'a V)>;
```

Comment:
I think this can be nicer if you have define a struct `Iter` that holds a `slice::Iter` or something like that and then impl the right traits for that `Iter` ... `Iterator`, `DoubleEndedIterator`, `ExactSizeIterator` ...

But maybe this isn't needed for now since it is all an experiment and can be improved upon later if it works out ...

Reply:
Yeah, sure this could certainly be improved (implement more traits etc.). If this really ends up being used more, it certainly makes sense to make this a little bit more advanced/add features (as you say with a wrapper type). But I think we should do that when we actually really need it (this rather simple addition already introduces quite a bit of LOCs).